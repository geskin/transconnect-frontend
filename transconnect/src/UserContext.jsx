import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

