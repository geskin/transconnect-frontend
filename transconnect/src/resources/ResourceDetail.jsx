import React, { useContext, useEffect, useState } from "react";
import { useNavigate, useParams } from "react-router-dom";
import TransconnectApi from "../api";
import UserContext from "../UserContext";

const ResourceDetail = () => {
    const params = useParams();
    console.debug(params);
    const id = params.id;
    const [resource, setResource] = useState({});
    const { currUser } = useContext(UserContext);
    const navigate = useNavigate();


    useEffect(() => {
        const fetchResource = async () => {
            try {
                const resource = await TransconnectApi.getResource(id);
                if (resource) {
                    setResource(resource);
                }
            } catch (err) {
                console.error("Error fetching resource", err);
            }
        }
        fetchResource();
    }, [id]);

    if (!resource.approved && currUser.role !== 'ADMIN') navigate('/resources');

    return (
        <div className="col-md-8 offset-md-2">
            {currUser.role === 'ADMIN' ?
                <div>
                    {resource.approved ? <i>Approved</i> : <i>Pending approval</i>}
                    <h3 className="mb-3 white-letters"><Link to={resource.url}>{resource.name}</Link></h3>
                    <p className="white-letters">{resource.description}</p>
                    <div>
                        {resource.types.map(t => (<i>{t.name}</i>))}
                    </div>
                    <div>
                        <button><Link to={`/resources/${resource.id}/edit`}>Edit</Link></button>
                        <Link to="/resources">Back to resources</Link>
                    </div>
                </div>
                :
                <div>
                    <h3 className="mb-3 white-letters"><Link to={resource.url}>{resource.name}</Link></h3>
                    <p className="white-letters">{resource.description}</p>
                    <div>
                        {resource.types.map(t => (<i>{t.name}</i>))}
                    </div>
                    <div>
                        <Link to="/resources">Back to resources</Link>
                    </div>
                </div>
            }
        </div>
    );
}

export default ResourceDetail;