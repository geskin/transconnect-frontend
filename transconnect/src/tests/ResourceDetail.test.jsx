import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import ResourceDetail from "../resources/ResourceDetail";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import { MemoryRouter, Route, Routes } from "react-router-dom";

vi.mock("../api");

describe("ResourceDetail", () => {
    const mockCurrUser = { username: "testuser", id: 1, role: "USER" };
    const mockResource = {
        id: 101,
        name: "Test Resource",
        description: "This is a test resource.",
        url: "http://test.com",
        types: [{ name: "Hotline" }, { name: "Other" }],
        approved: false,
    };

    beforeEach(() => {
        TransconnectApi.getResource.mockResolvedValue(mockResource);
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter initialEntries={["/resources/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources/:id" element={<ResourceDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter initialEntries={["/resources/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources/:id" element={<ResourceDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("fetches and displays resource details", async () => {
        render(
            <MemoryRouter initialEntries={["/resources/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources/:id" element={<ResourceDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getAllByText("Test Resource").length).toBeGreaterThan(0);
        });
    });

    it("displays resource types correctly", async () => {
        render(
            <MemoryRouter initialEntries={["/resources/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources/:id" element={<ResourceDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getAllByText("Hotline")).to.exist;
            expect(screen.getAllByText("Other")).to.exist;
        });
    });

    // it("handles resource approval toggle for admins", async () => {
    //     const mockAdminUser = { username: "admin", id: 2, role: "ADMIN" };
    //     const mockApprovedResource = { ...mockResource, approved: true };
    //     TransconnectApi.approve.mockResolvedValue(mockApprovedResource);

    //     render(
    //         <MemoryRouter initialEntries={["/resources/101"]}>
    //             <UserContext.Provider value={{ currUser: mockAdminUser }}>
    //                 <Routes>
    //                     <Route path="/resources/:id" element={<ResourceDetail />} />
    //                 </Routes>
    //             </UserContext.Provider>
    //         </MemoryRouter>
    //     );

    //     await waitFor(() => {
    //         expect(screen.getAllByText("Test Resource")).to.exist;
    //     });

    //     fireEvent.click(screen.getByText("Approve"));

    //     await waitFor(() => {
    //         expect(TransconnectApi.approve).toHaveBeenCalledWith(true, "101");
    //     });
    // });

    // it("handles resource deletion", async () => {
    //     TransconnectApi.deleteResource.mockResolvedValue({});

    //     render(
    //         <MemoryRouter initialEntries={["/resources/101"]}>
    //             <UserContext.Provider value={{ currUser: mockCurrUser }}>
    //                 <Routes>
    //                     <Route path="/resources/:id" element={<ResourceDetail />} />
    //                 </Routes>
    //             </UserContext.Provider>
    //         </MemoryRouter>
    //     );

    //     await waitFor(() => {
    //         expect(screen.getByText("Test Resource")).toBeInTheDocument();
    //     });

    //     fireEvent.click(screen.getByText("Delete"));

    //     await waitFor(() => {
    //         expect(TransconnectApi.deleteResource).toHaveBeenCalledWith("101");
    //     });
    // });

    it("redirects to /resources when user does not have admin role", async () => {
        render(
            <MemoryRouter initialEntries={["/resources/101"]}>
                <UserContext.Provider value={{ currUser: mockCurrUser }}>
                    <Routes>
                        <Route path="/resources/:id" element={<ResourceDetail />} />
                    </Routes>
                </UserContext.Provider>
            </MemoryRouter>
        );

        await waitFor(() => {
            expect(screen.getAllByText("Test Resource")).to.exist;
        });

        expect(screen.queryByText("Approve")).to.not.exist;
        expect(screen.queryByText("Delete")).to.not.exist;
    });
});
