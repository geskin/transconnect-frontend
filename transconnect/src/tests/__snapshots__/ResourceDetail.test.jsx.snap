// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceDetail > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root Card card css-ybsrxv-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiCardContent-root card-body css-1lt5qva-MuiCardContent-root"
    >
      <h2
        class="card-title text-start"
      >
        <b />
      </h2>
      <p
        class="text-start"
      >
        <i />
      </p>
      <div />
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1kdzw2s-MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Back to Resources
      </button>
    </div>
  </div>
</DocumentFragment>
`;
